### DISCUSSION TITLE
Refactor the representation of HTTP status codes & text

### DISCUSSION THREAD

**Discussion by webbeef** [servo/servo#33578] [General]
Generally speaking, a HTTP status code is a u16 in the 100..599 range, and a status message which is a byte sequence (https://fetch.spec.whatwg.org/#concept-response-status-message). Because of the requirement for the creation of responses representing a network error, the status can in these cases take a value of 0 (https://fetch.spec.whatwg.org/#concept-network-error).

To support that, we represent "HTTP status" in different ways:

In `script/dom/response.rs` : https://github.com/servo/servo/blob/a56c837122e76b243e7fadf3b48215af23f199c6/components/script/dom/response.rs#L43
Nothing ensures that `status` and `raw_status` are actually in kept in sync.

In `shared/net/response.rs` : https://github.com/servo/servo/blob/a56c837122e76b243e7fadf3b48215af23f199c6/components/shared/net/response.rs#L100
Here again, nothing ensures that `status` and `raw_status` are actually in kept in sync.

In `shared/net/lib.rs`:
https://github.com/servo/servo/blob/a56c837122e76b243e7fadf3b48215af23f199c6/components/shared/net/lib.rs#L75 and https://github.com/servo/servo/blob/a56c837122e76b243e7fadf3b48215af23f199c6/components/shared/net/lib.rs#L661

I'd like to harmonize all these in a single type that can address all these use cases and removes the need for manual synchronization of the `status` and `raw_status` that is (sometimes, but not always!) done like in https://github.com/servo/servo/blob/a56c837122e76b243e7fadf3b48215af23f199c6/components/net/protocols/mod.rs#L86

I think we can use a simple struct with private `u16` and `Vec<u8>` members, and appropriate methods to enforce the invariants we need.

Thoughts?

**Comment by jdm** [comment]
This sounds like a useful improvement!

**Comment by webbeef** [comment]
I submitted a PR: https://github.com/servo/servo/pull/33581

I'm quite happy with the ergonomic improvements, eg. for people implementing protocol handlers.

**Comment by webbeef** [comment]
Done in https://github.com/servo/servo/commit/58f34ad7a3db8c633871d568bfcef8b094217e3e
